// @microsoft/agents-hosting-extensions-teams@1.2.2/dist/src/messageExtension/messagingExtensionQuery.js
"use strict";
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.messagingExtensionQueryZodSchema = void 0;
exports.parseValueMessagingExtensionQuery = parseValueMessagingExtensionQuery;
const zod_1 = require("zod");
const messagingExtensionParameter_1 = require("./messagingExtensionParameter");
const messagingExtensionQueryOptions_1 = require("./messagingExtensionQueryOptions");
/**
 * Zod schema for validating MessagingExtensionQuery.
 */
exports.messagingExtensionQueryZodSchema = zod_1.z.object({
    commandId: zod_1.z.string().min(1).optional(),
    parameters: zod_1.z.array(messagingExtensionParameter_1.messagingExtensionParameterZodSchema).optional(),
    queryOptions: messagingExtensionQueryOptions_1.messagingExtensionQueryOptionsZodSchema.optional(),
    state: zod_1.z.string().min(1).optional()
});
/**
 * Parses the given value as a messaging extension query.
 *
 * @param {unknown} value - The value to parse.
 * @returns {MessagingExtensionQuery} - The parsed messaging extension query.
 */
function parseValueMessagingExtensionQuery(value) {
    exports.messagingExtensionQueryZodSchema.passthrough().parse(value);
    return value;
}
